### PR TITLE
refactor(ivy): use styling direct mode for all non-duplicate entries

### PR DESCRIPTION
**Pull request by matsko**
This patch ensures that all template-level style and class bindings
are processed using the direct-mode when their values are applied to
an element.

### CODE REVIEW

**Comment by matsko** [comment]
presubmit = https://tap.corp.google.com/ui#id=OCL:281647248:BASE:281647298:1574302010861:db78c133

ivy presubmit = https://test.corp.google.com/ui#id=OCL:281647248:BASE:281648237:1574302398351:ea091053

File: packages/compiler/src/render3/view/styling_builder.ts [hunk 1]
```diff
@@ -19,6 +19,7 @@ import {ValueConverter} from './template';
 import {getInterpolationArgsLength} from './util';
 
 const IMPORTANT_FLAG = '!important';
+export const MIN_STYLING_BINDING_SLOTS_REQUIRED = 2;
```
Comment (mhevery) [inline packages/compiler/src/render3/view/styling_builder.ts]: Please add documentation explaining why this is needed, and how it is used.

File: packages/compiler/src/render3/view/styling_builder.ts [hunk 2]
```diff
@@ -375,7 +376,12 @@ export class StylingBuilder {
           instructions[instructions.length - 1];
       const value = input.value.visit(valueConverter);
       let referenceForCall = reference;
-      let totalBindingSlotsRequired = 1;  // each styling binding value is stored in the LView
+
+      // each styling binding value is stored in the LView
+      // but there are two values stored for each binding:
+      //   1) the value itself
+      //   2) any intermediate values used for the direct-write algorithm
```
Comment (mhevery) [inline packages/compiler/src/render3/view/styling_builder.ts]:
```suggestion
      //   2) an intermediate value (concatenation of style up to this point).
      //      We need to store the intermediate value so that we don't allocate the strings on each CD.
```

File: packages/core/src/render3/instructions/element.ts [hunk 3]
```diff
@@ -64,7 +64,7 @@ export function ɵɵelementStart(
     }
   }
 
-  if ((tNode.flags & TNodeFlags.hasInitialStyling) === TNodeFlags.hasInitialStyling) {
+  if (tNode.flags & TNodeFlags.hasInitialStyling) {
```
Comment (mhevery) [inline packages/core/src/render3/instructions/element.ts]: Why this change? I think the original was correct?

File: packages/core/src/render3/interfaces/node.ts [hunk 4]
```diff
@@ -78,29 +78,37 @@ export const enum TNodeFlags {
    */
   hasHostBindings = 0x80,
 
-  /** Bit #9 - This bit is set if the node has initial styling */
-  hasInitialStyling = 0x100,
+  /** Bit #9 - This bit is set if the node has any initial classes */
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: Please add comment which provides value over just repeating what the code already says. A comment should answer:
- Why do we need this.
- How does this work on high level

And ideally the comment should contain an example which demonstrates why this is needed.

File: packages/core/src/render3/interfaces/node.ts [hunk 5]
```diff
@@ -78,29 +78,37 @@ export const enum TNodeFlags {
...
+  hasInitialClasses = 0x100,
+
+  /** Bit #10 - This bit is set if the node has any initial styles */
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: same here

File: packages/core/src/render3/interfaces/node.ts [hunk 6]
```diff
@@ -78,29 +78,37 @@ export const enum TNodeFlags {
...
+  hasInitialStyles = 0x200,
+
+  /**
+   * whether or not there are any initial styles or initial classes present
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: same here

File: packages/core/src/render3/interfaces/node.ts [hunk 7]
```diff
@@ -148,40 +156,50 @@ export const enum TNodeFlags {
...
    * Examples include:
    * 1. prop + prop: `<div [style.width]="x" dir-that-sets-width>`
    * 2. map + prop: `<div [style]="x" [style.prop]>`
    * 3. map + map: `<div [style]="x" dir-that-sets-style>`
    */
-  hasDuplicateStyleBindings = 0x40000,
+  hasDuplicateStyleBindings = 0x80000,
+
+  /**
+   * whether or not a `TStylingContext` instance is used for classes (merge mode algorithm active)
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: improve comment

File: packages/core/src/render3/interfaces/node.ts [hunk 8]
```diff
@@ -148,40 +156,50 @@ export const enum TNodeFlags {
...
+   */
+  hasClassContextInUse = 0x100000,
+
+  /**
+   * whether or not a `TStylingContext` instance is used for styles (merge mode algorithm active)
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: improve comment

File: packages/core/src/render3/interfaces/node.ts [hunk 9]
```diff
@@ -580,6 +598,11 @@ export interface TNode {
    */
   styles: StylingMapArray|TStylingContext|null;
 
+  /**
+   * Last binding index for any `[style]` or `[style.prop]` bindings on this node.
+   */
+  stylesBindingIndex: number;
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: - Why is this needed (what would not be possible or difficult to do otherwise)
- Add an example which demonstrates how this is used.

File: packages/core/src/render3/interfaces/node.ts [hunk 10]
```diff
@@ -599,6 +622,11 @@ export interface TNode {
    * will be placed into the initial styling slot in the newly created `TStylingContext`.
    */
   classes: StylingMapArray|TStylingContext|null;
+
+  /**
+   * Last binding index for any `[class]` or `[class.name]` bindings on this node.
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: here too. OK to reference style comment and just say same but for classes.

File: packages/core/src/render3/interfaces/styling.ts [hunk 11]
```diff
@@ -470,4 +470,18 @@ export const enum StylingMapsSyncMode {
  *
  * The styling algorithm code only needs access to `flags`.
  */
-export interface TStylingNode { flags: TNodeFlags; }
+export interface TStylingNode {
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/styling.ts]: This seems to be just a subset of `TNode`. If that is the case:
- what is the value in having it? (If it is really needed, can we beef up comments?)
- The type system should ensure that the two don't go out of sync. (So probably a check which asserts that one can be assigned to other?)

File: packages/core/src/render3/interfaces/styling.ts [hunk 12]
```diff
@@ -470,4 +470,18 @@ export const enum StylingMapsSyncMode {
...
+  classesBindingIndex: number;
+  stylesBindingIndex: number;
+  flags: TNodeFlags;
+}
+
+/**
+ * Various flags used for each style/class binding entry stored inside of `TData`.
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/styling.ts]: - Please add comment which would explain how this is used, why needed etc...
- Each flag should be documented likewise. Under what circumstances is the flag used, why we can't do without it.

**Review by mhevery — CHANGES_REQUESTED** [review]
I was not able to finish the review tonight, will look at it more tomorrow.

File: packages/core/src/render3/instructions/element.ts (hunk 3 above)
Comment (matsko) [reply packages/core/src/render3/instructions/element.ts]: The flag is now a combination of two other flags. Therefore, when either bit is set, then it should be true. This means the statement above fails unless it just checks for the presence of one of the true bits.

Comment (mhevery) [reply packages/core/src/render3/instructions/element.ts]: I see

File: packages/core/src/render3/interfaces/node.ts (hunk 4 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

File: packages/core/src/render3/interfaces/styling.ts (hunk 12 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/styling.ts]: done

File: packages/core/src/render3/interfaces/styling.ts (hunk 11 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/styling.ts]: done

File: packages/core/src/render3/interfaces/node.ts (hunk 10 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

File: packages/core/src/render3/interfaces/node.ts (hunk 9 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

File: packages/core/src/render3/interfaces/node.ts (hunk 8 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

File: packages/core/src/render3/interfaces/node.ts (hunk 7 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

File: packages/core/src/render3/interfaces/node.ts (hunk 6 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

File: packages/core/src/render3/interfaces/node.ts (hunk 5 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

File: packages/compiler/src/render3/view/styling_builder.ts (hunk 2 above)
Comment (matsko) [reply packages/compiler/src/render3/view/styling_builder.ts]: fixed

File: packages/compiler/src/render3/view/styling_builder.ts (hunk 1 above)
Comment (matsko) [reply packages/compiler/src/render3/view/styling_builder.ts]: done

File: packages/core/src/render3/instructions/styling.ts [hunk 13]
```diff
@@ -618,3 +643,66 @@ function patchHostStylingFlag(tNode: TNode, hostBindingsMode: boolean, isClassBa
...
+      i += StylingMapArrayIndex.TupleSize;
+    }
+  }
+
+  setPreviousBindingIndex(
+      tData, bindingIndex, previousBindingIndex, sanitizationRequired, hasPotentialOverlap);
+}
+
+function hyphenate(value: string): string {
+  return value.replace(/[a-z][A-Z]/g, v => v.charAt(0) + '-' + v.charAt(1)).toLowerCase();
```
Comment (mhevery) [inline packages/core/src/render3/instructions/styling.ts]: Extra regexp into `const` so that we don't allocate new regexp on each execution.

File: packages/core/src/render3/instructions/styling.ts [hunk 14]
```diff
@@ -618,3 +643,66 @@ function patchHostStylingFlag(tNode: TNode, hostBindingsMode: boolean, isClassBa
...
+ */
+export function registerBindingIntoTNode(
+    tNode: TNode, tData: TData, prop: string | null, bindingIndex: number, isClassBased: boolean,
+    sanitizationRequired: boolean): void {
+  tData[bindingIndex] = (isClassBased && prop !== null) ? hyphenate(prop) : prop;
+
+  let previousBindingIndex: number;
+  if (isClassBased) {
+    previousBindingIndex = tNode.classesBindingIndex;
+    tNode.classesBindingIndex = bindingIndex;
```
Comment (mhevery) [inline packages/core/src/render3/instructions/styling.ts]: It seems like `classesBindingIndex` points to tail. This is not obvious from the name of property. I also think that updating `TNode` like this is an issue. It means that if we reading a binding would have side-effects we could re-enter this code, and we could have two places where such update would happen. (In other words this code is not re-entry safe)

I think `classesBindingIndex` should always point to the beginning of the chain. Let's discuss this.

File: packages/core/src/render3/interfaces/node.ts [hunk 15]
```diff
@@ -78,29 +78,63 @@ export const enum TNodeFlags {
...
-  /** Bit #9 - This bit is set if the node has initial styling */
-  hasInitialStyling = 0x100,
+  /**
+   * Bit #9 - This bit is set if the node has any initial classes
+   *
+   * Tracking initial classes using this flag is important for performance
+   * reasons because it enables the class binding code to avoid certain
+   * operations (i.e. if a class value doesn't exist both in binding code
+   * and in the initial classes code then there is no need to search/replace
+   * the class out of the className string).
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: code example would make this even better.

File: packages/core/src/render3/interfaces/node.ts [hunk 16]
```diff
@@ -78,29 +78,63 @@ export const enum TNodeFlags {
...
+   * and in the initial styles code then there is no need to search/replace
+   * the class out of the style attribute string).
+   */
+  hasInitialStyles = 0x200,
+
+  /**
+   * Whether or not there are any initial styles or initial classes present
+   *
+   * This flag is a combination of both flags for tracking initial styles
+   * and classes. There are various cases in element creation that the
```
Comment (mhevery) [inline packages/core/src/render3/interfaces/node.ts]: `There are various cases ` List the cases under which this happens as bullet list. The point of documentation is to explain exactly why it is needed.

**Review by mhevery — CHANGES_REQUESTED** [review]
More comments, will continue after lunch

File: packages/compiler/src/render3/view/compiler.ts [hunk 17]
```diff
@@ -547,10 +546,36 @@ function createHostBindingsFunction(
...
+
+  // figure out how many min host vars are required before
+  // interpolations and values are converted.
+  let totalHostVarsCount = 0;
+  bindings && bindings.forEach((binding: ParsedProperty) => {
+    const name = binding.name;
+    const stylingInputWasSet =
+        styleBuilder.registerInputBasedOnName(name, binding.expression, binding.sourceSpan);
+    if (stylingInputWasSet) {
+      totalHostVarsCount += MIN_STYLING_BINDING_SLOTS_REQUIRED;
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/compiler/src/render3/view/compiler.ts]: We use `MIN` because interpolation adds additional slot requirements on top of the minimum entries.

File: packages/compiler/src/render3/view/compiler.ts [hunk 18]
```diff
@@ -671,7 +680,8 @@ function createHostBindingsFunction(
         instruction.calls.forEach(call => {
           // we subtract a value of `1` here because the binding slot was already allocated
           // at the top of this method when all the input bindings were counted.
-          totalHostVarsCount += Math.max(call.allocateBindingSlots - 1, 0);
+          totalHostVarsCount +=
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/compiler/src/render3/view/compiler.ts]: We first increment the total binding amounts necessary before pipes and interpolation values are added. This is the nature of host bindings code, because all other bindings need to know exactly how many slots are required before they are evaluated. This is why we increment the style/class bindings first (using that constant) and then subtract them out later one (when the interpolations are evaluated).

File: packages/core/src/render3/instructions/styling.ts [hunk 19]
```diff
@@ -175,14 +182,11 @@ export function ɵɵclassProp(className: string, value: boolean | null): Tsickle
...
+    tNode: TNode, firstUpdatePass: boolean, tData: TData, lView: LView, bindingIndex: number,
+    prop: string, value: boolean | number | SafeValue | string | null | undefined | NO_CHANGE,
     isClassBased: boolean): boolean {
   let updated = false;
-
   const native = getNativeByTNode(tNode, lView) as RElement;
-  const context = isClassBased ? getClassesContext(tNode) : getStylesContext(tNode);
-  const sanitizer = isClassBased ? null : getCurrentStyleSanitizer();
 
   // [style.prop] and [class.name] bindings do not use `bind()` and will
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/core/src/render3/instructions/styling.ts]: We do this outside of the instruction with our own binding logic (the `styleProp`, `classProp`, `styleMap` and `classMap` instructions all do their own reading/writing to `LView`).

File: packages/core/src/render3/instructions/styling.ts (hunk 13 above)
Comment (matsko) [reply packages/core/src/render3/instructions/styling.ts]: fixed

File: packages/core/src/render3/instructions/styling.ts (hunk 14 above)
Comment (matsko) [reply packages/core/src/render3/instructions/styling.ts]: The problem is that host bindings have no way of connecting back to template bindings if we do a head-only approach. 

The `tNode.classBindingIndex` and `tNode.stylesBindingIndex` values only get updated during the `firstUpdatePass` for a single instance of the first element. Then each member stays the same for future updates. We do the same thing thing with `TStlyingContext` as new bindings are registered.

File: packages/core/src/render3/interfaces/node.ts (hunk 15 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: done

File: packages/core/src/render3/interfaces/node.ts (hunk 16 above)
Comment (matsko) [reply packages/core/src/render3/interfaces/node.ts]: fixed

File: packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts [hunk 20]
```diff
@@ -504,7 +504,7 @@ describe('compiler compliance: styling', () => {
               type: MyComponent,
               selectors:[["my-component"]],
               decls: 1,
-              vars: 5,
+              vars: 7,
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts]: That is just a regular binding. So nothing happens. We add `+2` here because of the two style.prop bindings (`[style.width]` and `[style.height]`). We don't add `+3` because `[style]` already uses two binding slots prior to this PR.

File: packages/compiler/src/render3/view/compiler.ts [hunk 21]
```diff
@@ -547,10 +546,36 @@ function createHostBindingsFunction(
...
+  if (eventBindings && eventBindings.length) {
+    const listeners = createHostListeners(eventBindings, name);
+    createStatements.push(...listeners);
+  }
+
+  // Calculate the host property bindings
+  const bindings = bindingParser.createBoundHostProperties(directiveSummary, hostBindingSourceSpan);
+  const allOtherBindings: ParsedProperty[] = [];
+
+  // figure out how many min host vars are required before
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/compiler/src/render3/view/compiler.ts]: fixed.

File: packages/compiler/src/render3/view/compiler.ts [hunk 22]
```diff
@@ -531,7 +531,6 @@ function createHostBindingsFunction(
     bindingParser: BindingParser, constantPool: ConstantPool, selector: string,
     name?: string): o.Expression|null {
   // Initialize hostVarsCount to number of bound host properties (interpolations illegal)
```
[the remark this answers is not in the document]
Comment (matsko) [reply packages/compiler/src/render3/view/compiler.ts]: fixed. thank you.

File: packages/core/src/render3/styling/direct_write_algorithm.ts [hunk 23]
```diff
@@ -0,0 +1,327 @@
...
+ * @returns `true` when the binding has updated (due to a value change).
+ */
+export function setStylingValue(
+    lView: LView, tNode: TNode, tData: TData, state: StylingState, value: any, bindingIndex: number,
+    sanitizer: StyleSanitizeFn | null, isClassBased: boolean): boolean {
+  const lastUpdatedBindingIndex =
+      isClassBased ? state.lastClassBindingIndex : state.lastStyleBindingIndex;
+  const previousValue = getValue(lView, bindingIndex);
+  const updated = hasValueChanged(previousValue, value);
+  if (lastUpdatedBindingIndex < 0 || updated) {
```
Comment (mhevery) [inline packages/core/src/render3/styling/direct_write_algorithm.ts]: Why is the check for value changed so deep in the algorithm? I would assume that it would simpler if it would be close to the `styleProp` (and friends) function, so that in the case when there is no change we could exit quickly.

File: packages/core/src/render3/styling/state.ts [hunk 24]
```diff
@@ -59,25 +58,11 @@ export interface StylingState {
...
-  /**
-   * The last class map that was applied (i.e. `[class]="x"`).
-   *
-   * Note that this property is only populated when direct class values are applied
-   * (i.e. context resolution is not used).
-   *
-   * See `allowDirectStyling` for more info.
-  */
-  lastDirectClassMap: StylingMapArray|null;
+  lastClassBindingIndex: number;
```
Comment (mhevery) [inline packages/core/src/render3/styling/state.ts]: Could we have docs explaining how each of these properties are used.

File: packages/core/src/render3/util/styling_utils.ts [hunk 25]
```diff
@@ -455,3 +449,192 @@ export function splitOnWhitespace(text: string): string[]|null {
 export function selectClassBasedInputName(inputs: PropertyAliases): string {
   return inputs.hasOwnProperty('class') ? 'class' : 'className';
 }
+
+/**
+ * Returns a style entry with the `prop` and `value` concatenated together
+ */
+export function concatStyle(prop: string, value: string) {
+  return `${prop}${STYLE_PROP_VALUE_SEPARATOR}${value}`;
```
Comment (mhevery) [inline packages/core/src/render3/util/styling_utils.ts]:
```suggestion
  return prop +STYLE_PROP_VALUE_SEPARATOR + value;
```
Is a lot more efficient

**Review by mhevery — CHANGES_REQUESTED** [review]
